File: CODE/DART/arc_vertex.dart
import 'arc_backward_vector_type.dart';
import 'arc_vector_3.dart';

// -- TYPES

class ArcVertex
{
    // -- ATTRIBUTES

    ArcVector3
        positionVector,
        forwardVector;
    late ArcVector3
        backwardVector;
    ArcBackwardVectorType
        backwardVectorType;

    // -- CONSTRUCTORS

    ArcVertex(
        {
            required this.positionVector,
            required this.forwardVector,
            ArcVector3? backwardVector,
            this.backwardVectorType = ArcBackwardVectorType.custom
        }
        )
    {
        if ( backwardVectorType == ArcBackwardVectorType.custom )
        {
            this.backwardVector =
                backwardVector ??
                ArcVector3(
                    x: 0,
                    y: 0,
                    z: 0,
                    );
        }
        else if ( backwardVectorType == ArcBackwardVectorType.symmetrical )
        {
            this.backwardVector =
                ArcVector3(
                    x: -forwardVector.x,
                    y: -forwardVector.y,
                    z: -forwardVector.z,
                    );
        }
        else
        {
            this.backwardVector =
                ArcVector3(
                    x: 0,
                    y: 0,
                    z: 0,
                    );
        }
    }

    // -- OPERATIONS

    void setForwardPositionVector(
        ArcVector3 forwardPositionVector
        )
    {
        forwardPositionVector.x = positionVector.x + forwardVector.x;
        forwardPositionVector.y = positionVector.y + forwardVector.y;
        forwardPositionVector.z = positionVector.z + forwardVector.z;
    }

    // ~~

    void setBackwardPositionVector(
        ArcVector3 backwardPositionVector
        )
    {
        backwardPositionVector.x = positionVector.x + backwardVector.x;
        backwardPositionVector.y = positionVector.y + backwardVector.y;
        backwardPositionVector.z = positionVector.z + backwardVector.z;
    }

    // ~~

    void setSymmetricalBackwardVector(
        )
    {
        backwardVector.x = -forwardVector.x;
        backwardVector.y = -forwardVector.y;
        backwardVector.z = -forwardVector.z;
    }

    // ~~

    void setResidualBackwardVector(
        ArcVertex firstVertex
        )
    {
        backwardVector.x = firstVertex.positionVector.x + firstVertex.forwardVector.x - positionVector.x;
        backwardVector.y = firstVertex.positionVector.y + firstVertex.forwardVector.y - positionVector.y;
        backwardVector.z = firstVertex.positionVector.z + firstVertex.forwardVector.z - positionVector.z;
    }
}

File: CODE/PYTHON/arc_vertex.py
# -- IMPORTS

from __future__ import annotations;
from typing import Optional;
from arc_backward_vector_type import ArcBackwardVectorType;
from arc_vector_3 import ArcVector3;

# -- TYPES

class ArcVertex:

    # -- CONSTRUCTORS

    def __init__(
        self,
        positionVector: ArcVector3,
        forwardVector: ArcVector3,
        backwardVector: Optional[ ArcVector3 ] = None,
        backwardVectorType: ArcBackwardVectorType = ArcBackwardVectorType.custom
        ) -> None:

        self.positionVector = positionVector;
        self.forwardVector = forwardVector;
        self.backwardVectorType = backwardVectorType;

        if ( backwardVectorType == ArcBackwardVectorType.custom ):
            self.backwardVector = backwardVector if backwardVector else ArcVector3( 0, 0, 0 );
        elif ( backwardVectorType == ArcBackwardVectorType.symmetrical ):
            self.backwardVector = ArcVector3( -forwardVector.x, -forwardVector.y, -forwardVector.z );

        else:
            self.backwardVector = ArcVector3( 0, 0, 0 );

    # -- OPERATIONS

    def setForwardPositionVector(
        self,
        forwardPositionVector: ArcVector3
        ) -> None:

        forwardPositionVector.x = self.positionVector.x + self.forwardVector.x;
        forwardPositionVector.y = self.positionVector.y + self.forwardVector.y;
        forwardPositionVector.z = self.positionVector.z + self.forwardVector.z;

    # ~~

    def setBackwardPositionVector(
        self,
        backwardPositionVector: ArcVector3
        ) -> None:

        backwardPositionVector.x = self.positionVector.x + self.backwardVector.x;
        backwardPositionVector.y = self.positionVector.y + self.backwardVector.y;
        backwardPositionVector.z = self.positionVector.z + self.backwardVector.z;

    # ~~

    def setSymmetricalBackwardVector(
        self
        ) -> None:

        self.backwardVector.x = -self.forwardVector.x;
        self.backwardVector.y = -self.forwardVector.y;
        self.backwardVector.z = -self.forwardVector.z;

    # ~~

    def setResidualBackwardVector(
        self,
        firstVertex: ArcVertex
        ) -> None:

        self.backwardVector.x = firstVertex.positionVector.x + firstVertex.forwardVector.x - self.positionVector.x;
        self.backwardVector.y = firstVertex.positionVector.y + firstVertex.forwardVector.y - self.positionVector.y;
        self.backwardVector.z = firstVertex.positionVector.z + firstVertex.forwardVector.z - self.positionVector.z;
